### DISCUSSION TITLE
Don't cast dataframe values to strings

### DISCUSSION THREAD

**Discussion by Spothedog1** [morph-kgc/morph-kgc#266] [Ideas]
The[ _preprocess_data](https://github.com/morph-kgc/morph-kgc/blob/29d4385046db1a88f3c82549561d1bf5da6e75ce/src/morph_kgc/materializer.py#L30) function converts all the values in the dataframe to strings. 

However, this is undesirable behavior when trying to unpack data structures from cell values.

I'm trying to materialize [OvertureMaps Divisions](https://docs.overturemaps.org/schema/reference/divisions/division/) data into an RDF Knowledge Graph. The files are stored in `.parquet` and many of the cells have structs or geometry values as their values. 

To deal with this I have been writing udfs, but the input to my udfs is always a string representation of the cell value, causing me to have to use `ast.literal_eval(param)` to unpack the value and get the necessary information out of it.

This has worked for simpler structs, but the more complex ones I've been unable to solve

Here's an example of unpacking the geometry column
```python
@udf(fun_id=f"{prefix}loadGeometry", geometry=f"{prefix}geometry")
def load_geometry(geometry: str) -> str:
    """
    Function takes a string representation of a binary blob and returns the geometry as a WKT string

    :param geometry: The string representation of the binary blob of the geometry
    :return: The geometry as a WKT string
    """

    # Turn the geometry string into bytes and create a dataframe out of it
    df = pd.DataFrame({"geometry": [ast.literal_eval(geometry)]})
    return_df = duckdb.sql(f"SELECT ST_AsText(ST_GeomFromWKB(geometry::BLOB)) AS geometry FROM df").df()

    return return_df.iloc[0].geometry
```

However, for the columns `local_type` and `names` the structures get complex. Here are examples of strings being passed into my udfs. Using `ast.literal_eval` doesn't work for these.

```
array_str = "[array(['en', 'neighborhood'], dtype=object)]"
dictionary_str = "{'common': array([array(['en', 'Bellevue'], dtype=object)], dtype=object), 'primary': 'Bellevue', 'rules': None}"
multiple_elements_str = "array([{'between': None, 'language': None, 'side': None, 'value': 'The Hill', 'variant': 'short'}, {'between': None, 'language': 'es', 'side': None, 'value': 'Distrito de Capitol Hill', 'variant': 'alternate'}], dtype=object)"
```

When they're initially loaded into the dataframe by the [read_parquet](https://github.com/morph-kgc/morph-kgc/blob/29d4385046db1a88f3c82549561d1bf5da6e75ce/src/morph_kgc/data_source/data_file.py#L88) function, they're stored as objects that can easily be worked with and have the relevant data extracted from. I've tested this by running my udfs on the [dataframe](https://github.com/morph-kgc/morph-kgc/blob/29d4385046db1a88f3c82549561d1bf5da6e75ce/src/morph_kgc/materializer.py#L61) before the preprocess step. 

What are your thoughts on not casting all values to strings and leaving them as their original pandas types when an fnml function is going to be executed on the value?

In my example above they are simple dictionaries and numpy arrays, it's going to take a small custom library to parse all those string variants into their correct datatype so I can extract the value that I need.

**Comment by arenas-guerrero-julian** [comment]
This is an quite relevant point. Indeed, I though about this some time ago but in the end I maintained the string casting behaviour. A similar issue occurs when working with numbers, you need to cast them back to int or float before doing math operations.

We can take a look into this in more detail.

**Reply by Spothedog1** [reply]
1. If we made this change would it only be on the condition that a function is going to be executed on the dataframe?
2. What downstream affects would we have to handle?

**Reply by arenas-guerrero-julian** [reply]
The problem is that there are some string-ish operation that are required for the generation of RDF, such as scaping and percent encoding. A solution to explore is  delaying casting to string and performing these operations after RML-FNML processing, but we need to check if that is actually possible.
